# HG changeset patch
# User Torsten <mttl@mailbox.org>
# Date 1545638121 -3600
#      Mon Dec 24 08:55:21 2018 +0100
# Branch stable
# Node ID c95884eed1cf3a9b1e5f17d6e68679dd3fdd61f7
# Parent  9254cdba20b269f76f913acc6680a18f80041e71
make restoring gui more robust to changes of desktop size (bug #53207)

* default-qt-settings.in: remove remaining entries for main window

* gui-preferences.h: define keys and defaults for main window state and geometry

* main-window.cc (set_window_layout): Check for successful restoring state and
  geometry and take default state and geometry in case of error

* octave-dock-widget.cc: include QDesktopWidget,
  (handle_settings): define default geometry from the available desktop size
  and also use this default if the geometry from the settings file is not
  included within the available desktop size

diff --git a/libgui/default-qt-settings.in b/libgui/default-qt-settings.in
--- a/libgui/default-qt-settings.in
+++ b/libgui/default-qt-settings.in
@@ -20,20 +20,16 @@ restoreSession=false
 savedSessionTabs=@Invalid()
 mru_file_list=@Invalid()
 
 [terminal]
 fontSize=__default_font_size__
 fontName=__default_font__
 cursorType=ibeam
 
-[MainWindow]
-geometry=@ByteArray(\x1\xd9\xd0\xcb\0\x2\0\0\0\0\0\0\0\0\0\x1b\0\0\x4\x3\0\0\x3\t\0\0\0\0\0\0\0\x39\0\0\x4\x3\0\0\x3\t\0\0\0\0\0\0\0\0\x5\0)
-windowState="@ByteArray(\0\0\0\xff\0\0\0\0\xfd\0\0\0\x1\0\0\0\x1\0\0\x4\x4\0\0\x2|\xfc\x2\0\0\0\x1\xfc\0\0\0?\0\0\x2|\0\0\x1\x98\0\xff\xff\xff\xfc\x1\0\0\0\x2\xfc\0\0\0\0\0\0\x1W\0\0\0\x9b\0\xff\xff\xff\xfc\x2\0\0\0\x3\xfb\0\0\0\x1e\0\x46\0i\0l\0\x65\0s\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0?\0\0\0\xf9\0\0\0\x88\0\xff\xff\xff\xfb\0\0\0\x1a\0W\0o\0r\0k\0s\0p\0\x61\0\x63\0\x65\0V\0i\0\x65\0w\x1\0\0\x1>\0\0\0\xcb\0\0\0\x82\0\xff\xff\xff\xfb\0\0\0\"\0H\0i\0s\0t\0o\0r\0y\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\x2\xf\0\0\0\xac\0\0\0\x82\0\xff\xff\xff\xfc\0\0\x1]\0\0\x2\xa7\0\0\x1O\0\xff\xff\xff\xfc\x2\0\0\0\x2\xfb\0\0\0\x1c\0N\0\x65\0w\0s\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0@\0\0\0\xdb\0\0\0\0\0\0\0\0\xfc\0\0\0?\0\0\x2|\0\0\0\xeb\x1\0\0\x1b\xfa\0\0\0\0\x2\0\0\0\x5\xfb\0\0\0$\0T\0\x65\0r\0m\0i\0n\0\x61\0l\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\x46\0\xff\xff\xff\xfb\0\0\0\x14\0\x46\0i\0l\0\x65\0\x45\0\x64\0i\0t\0o\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\x62\0\xff\xff\xff\xfb\0\0\0\x1c\0V\0\x61\0r\0i\0\x61\0\x62\0l\0\x65\0\x45\0\x64\0i\0t\0o\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0;\0\xff\xff\xff\xfb\0\0\0.\0\x44\0o\0\x63\0u\0m\0\x65\0n\0t\0\x61\0t\0i\0o\0n\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\xcf\0\xff\xff\xff\xfb\0\0\0\x1e\0v\0\x61\0r\0i\0\x61\0\x62\0l\0\x65\0_\0\x65\0\x64\0i\0t\0o\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x2|\0\0\0\x4\0\0\0\x4\0\0\0\b\0\0\0\b\xfc\0\0\0\x1\0\0\0\x2\0\0\0\x1\0\0\0\x16\0M\0\x61\0i\0n\0T\0o\0o\0l\0\x42\0\x61\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0)"
-
 [DockWidgets]
 TerminalDockWidget=@ByteArray(\x1\xd9\xd0\xcb\0\x1\0\0\0\0\x1j\0\0\0 \0\0\x4\x3\0\0\x2\x9e\0\0\x1j\0\0\0 \0\0\x4\x3\0\0\x2\x9e\0\0\0\0\0\0)
 TerminalDockWidgetFloating=false
 TerminalDockWidgetVisible=true
 WorkspaceView=@ByteArray(\x1\xd9\xd0\xcb\0\x1\0\0\0\0\0\0\0\0\x1+\0\0\x1\x63\0\0\x2\0\0\0\0\0\0\0\x1+\0\0\x1\x63\0\0\x2\0\0\0\0\0\0\0)
 WorkspaceViewFloating=false
 WorkspaceViewVisible=true
 HistoryDockWidget=@ByteArray(\x1\xd9\xd0\xcb\0\x1\0\0\0\0\0\0\0\0\x2\a\0\0\x1\x63\0\0\x2\xba\0\0\0\0\0\0\x2\a\0\0\x1\x63\0\0\x2\xba\0\0\0\0\0\0)
diff --git a/libgui/src/gui-preferences.h b/libgui/src/gui-preferences.h
--- a/libgui/src/gui-preferences.h
+++ b/libgui/src/gui-preferences.h
@@ -61,16 +61,24 @@ const QStyle::PixelMetric global_icon_si
   QStyle::PM_LargeIconSize
 };
 const gui_pref global_icon_size ("toolbar_icon_size", QVariant (0));
 const gui_pref global_icon_theme ("use_system_icon_theme", QVariant (true));
 
 // Style
 const gui_pref global_style ("style", QVariant ("default"));
 
+// Main window preferences
+
+const gui_pref mw_geometry ("MainWindow/geometry",
+  QVariant ("\x1\xd9\xd0\xcb\0\x2\0\0\0\0\0\0\0\0\0\x1b\0\0\x4\x3\0\0\x3\t\0\0\0\0\0\0\0\x39\0\0\x4\x3\0\0\x3\t\0\0\0\0\0\0\0\0\x5\0"));
+
+const gui_pref mw_state ("MainWindow/state",
+  QVariant ("\0\0\0\xff\0\0\0\0\xfd\0\0\0\x1\0\0\0\x1\0\0\x4\x4\0\0\x2|\xfc\x2\0\0\0\x1\xfc\0\0\0?\0\0\x2|\0\0\x1\x98\0\xff\xff\xff\xfc\x1\0\0\0\x2\xfc\0\0\0\0\0\0\x1W\0\0\0\x9b\0\xff\xff\xff\xfc\x2\0\0\0\x3\xfb\0\0\0\x1e\0\x46\0i\0l\0\x65\0s\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0?\0\0\0\xf9\0\0\0\x88\0\xff\xff\xff\xfb\0\0\0\x1a\0W\0o\0r\0k\0s\0p\0\x61\0\x63\0\x65\0V\0i\0\x65\0w\x1\0\0\x1>\0\0\0\xcb\0\0\0\x82\0\xff\xff\xff\xfb\0\0\0\"\0H\0i\0s\0t\0o\0r\0y\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\x2\xf\0\0\0\xac\0\0\0\x82\0\xff\xff\xff\xfc\0\0\x1]\0\0\x2\xa7\0\0\x1O\0\xff\xff\xff\xfc\x2\0\0\0\x2\xfb\0\0\0\x1c\0N\0\x65\0w\0s\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0@\0\0\0\xdb\0\0\0\0\0\0\0\0\xfc\0\0\0?\0\0\x2|\0\0\0\xeb\x1\0\0\x1b\xfa\0\0\0\0\x2\0\0\0\x5\xfb\0\0\0$\0T\0\x65\0r\0m\0i\0n\0\x61\0l\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\x46\0\xff\xff\xff\xfb\0\0\0\x14\0\x46\0i\0l\0\x65\0\x45\0\x64\0i\0t\0o\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\x62\0\xff\xff\xff\xfb\0\0\0\x1c\0V\0\x61\0r\0i\0\x61\0\x62\0l\0\x65\0\x45\0\x64\0i\0t\0o\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0;\0\xff\xff\xff\xfb\0\0\0.\0\x44\0o\0\x63\0u\0m\0\x65\0n\0t\0\x61\0t\0i\0o\0n\0\x44\0o\0\x63\0k\0W\0i\0\x64\0g\0\x65\0t\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\xcf\0\xff\xff\xff\xfb\0\0\0\x1e\0v\0\x61\0r\0i\0\x61\0\x62\0l\0\x65\0_\0\x65\0\x64\0i\0t\0o\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x2|\0\0\0\x4\0\0\0\x4\0\0\0\b\0\0\0\b\xfc\0\0\0\x1\0\0\0\x2\0\0\0\x1\0\0\0\x16\0M\0\x61\0i\0n\0T\0o\0o\0l\0\x42\0\x61\0r\x1\0\0\0\0\xff\xff\xff\xff\0\0\0\0\0\0\0\0"));
+
 // Console preferences
 
 const gui_pref cs_font ("terminal/fontName", QVariant ());
 
 
 // Variable Editor preferences
 
 const gui_pref ve_font ("variable_editor/font_size", QVariant ());
diff --git a/libgui/src/main-window.cc b/libgui/src/main-window.cc
--- a/libgui/src/main-window.cc
+++ b/libgui/src/main-window.cc
@@ -1276,18 +1276,25 @@ namespace octave
 
   void main_window::init_terminal_size (void)
   {
     emit init_terminal_size_signal ();
   }
 
   void main_window::set_window_layout (QSettings *settings)
   {
-    restoreState (settings->value ("MainWindow/windowState").toByteArray ());
-    restoreGeometry (settings->value ("MainWindow/geometry").toByteArray ());
+    // Restore main window state and geometry from settings file or, in case
+    // of an error from the default layout
+    if (! restoreState (
+            settings->value (mw_state.key, mw_state.def).toByteArray ()))
+      restoreState (mw_state.def.toByteArray ());
+
+    if (! restoreGeometry (
+            settings->value (mw_geometry.key, mw_geometry.def).toByteArray ()))
+      restoreGeometry (mw_geometry.def.toByteArray ());
 
     // Restore the geometry of all dock-widgets
     foreach (octave_dock_widget *widget, dock_widget_list ())
       {
         QString name = widget->objectName ();
 
         if (! name.isEmpty ())
           {
diff --git a/libgui/src/octave-dock-widget.cc b/libgui/src/octave-dock-widget.cc
--- a/libgui/src/octave-dock-widget.cc
+++ b/libgui/src/octave-dock-widget.cc
@@ -21,22 +21,23 @@ along with Octave; see the file COPYING.
 
 */
 
 #if defined (HAVE_CONFIG_H)
 #  include "config.h"
 #endif
 
 #include <QApplication>
-#include <QToolBar>
 #include <QAction>
+#include <QDesktopWidget>
 #include <QHBoxLayout>
 #include <QLabel>
 #include <QSettings>
 #include <QStyle>
+#include <QToolBar>
 
 #include "resource-manager.h"
 #include "octave-dock-widget.h"
 
 namespace octave
 {
   label_dock_widget::label_dock_widget (QWidget *p)
     : QDockWidget (p), m_default_float_button (nullptr),
@@ -437,19 +438,27 @@ namespace octave
       m_icon_color = "";
 
     bcola.getRgb (&r, &g, &b);
     if (r+g+b < 400)
       m_icon_color_active = "-light";
     else
       m_icon_color_active = "";
 
+    QRect available_size = QApplication::desktop ()->availableGeometry ();
+    int x1, y1, x2, y2;
+    available_size.getCoords (&x1, &y1, &x2, &y2);
+    QRect default_size = QRect (x1, y1, x2/3, y2/2);
+
     m_recent_float_geom = settings->value ("DockWidgets/" + objectName ()
                                            + "_floating_geometry",
-                                           QRect (50,100,480,480)).toRect ();
+                                           default_size).toRect ();
+
+    if (! available_size.contains (m_recent_float_geom, false))
+      m_recent_float_geom = default_size;
 
     m_recent_dock_geom = settings->value ("DockWidgets/" + objectName (),
                                           QByteArray ()).toByteArray ();
 
     notice_settings (settings);  // call individual handler
 
     set_style (false);
   }
